### ISSUE TITLE
multiprocess plugin hanging during test execution

### ISSUE THREAD

**Issue by jredl-va** [nose-devs/nose2#284]
We've encountered an issues lately that is quite confusing. Our CI server runs tests in parallel with 4 processes. However, it has started hanging every once in awhile. Some digging has allowed us to create an example TestSuite that highlights this behavior:

```python
import unittest

class ThreeTests(unittest.TestCase):

    def test_one(self):
        pass

    def test_two(self):
        pass

    def test_three(self):
        pass
```

running the following will show nose2 hang:
```bash
nose2 three_tests --plugin=nose2.plugins.mp -N 4
```

The hypothesis is that when running the tests in bulk (10,000+ tests in many suites), that a test suite with three tests is firing at times that is not initially sending an event to the process? 

We've done some playing with where the test suite is flattened and distributed but these seems to have consequences on overall test run times: https://github.com/nose-devs/nose2/blob/master/nose2/plugins/mp.py#L75

``` python
        # send one initial task to each process
        for proc, conn in procs:
            if not flat and self.procs_initialized is True:
                break
            if not flat and self.procs_initialized is False:
                conn.send(None)
                continue
            caseid = flat.pop(0)
            conn.send(caseid)
        self.procs_initialized = True
```

Anyone have any other ideas?

**Comment by jredl-va** [comment]
Tagging @ptthiem to see if you have any suggestions as you were the main contributor to this file?

**Comment by little-dude** [comment]
hi @jredl-va thanks for debugging it. I probably won't be able to work on it this week end, but I'll look into it next week if @ptthiem does not chime in in the meantime :)

**Comment by ptthiem** [comment]
On windows? Or Posix?

**Comment by jredl-va** [comment]
Replicated on Mac osx latest release. CI server is ubunto 14 something.

**Comment by ptthiem** [comment]
Thx, jredl-va.  At least we don't have windows to deal with :D.  

Yeah the workaround would shutdown any worker processes on start up if there are not enough tests to give at least one to each out initially.  I'm not sure why would impact run-times much because a process pool is constructed for each execution of _runmp, are not persisted, and _runmp hooks into startTestRun.  Wouldn't this get run only once per execution anyway (someone)?  I am likely just missing something, trying to remember.  In any case, the loop condition is rdrs or flat.  Looks like we continue while there are tests *or* there are alive processes.  Processes appear to removed only when they return None.  Which if they never get a test and don't fail in init, then they are happily blocking on conn.recv().  That is probably why it is blocking.

In the error handling of EOFError at line 88, at least, we probably should also check for (or assume) a dead process too and see that rdrs.remove(conn) is done.

**Comment by jredl-va** [comment]
Here is a screen capture with the debugger open. You will notice that we get into an infinite loop. select.select never returns the waiting connection handler  (I believe because nothing has ever been submitted to it). 

![image](https://cloud.githubusercontent.com/assets/2300103/13716449/bc42a956-e7a0-11e5-8ac9-f53f69f77e45.png)

**Comment by ptthiem** [comment]
Yes an infinite loop would result.  The children are blocked on conn.recv.  So they never get into "ready", yet they are never removed from "rdrs" because they were alive when it was constructed.

**Comment by ptthiem** [comment]
Well at least I believe that to be the case.

**Comment by ptthiem** [comment]
I can probably spend some time on this this weekend.  The only thing that concerns me at that is it sounds like closing those processes as you have done, is decreasing performance.  I was thinking to allocate min(len(flat), self.procs) processes in the first place.

**Comment by ptthiem** [comment]
The alternatives, seem more complicated to me, and I really wasn't seeing in overhead.  I suppose we could poll the socket/pipe on the worker side.

**Comment by little-dude** [comment]
Thanks @ptthiem :)
The PR looks good.
@jredl-va does the PR fix your issue? If yes I'm going to merge it and make a release with the fix.

**Comment by jredl-va** [comment]
@ptthiem thanks for the PR, much more elegant than the approach I was taking. @little-dude yes, this corrects the issue we were experiencing.

**Comment by little-dude** [comment]
@jredl-va I just released 0.6.4, it should be available on pypi within 15 minutes or so. Closing, but feel free to re-open if needed.
